cli/src/cli.rs

```diff
@@ -1131,24 +1146,65 @@ fn process_deploy(
         bpf_loader::id()
     };
 
-    // Build transactions to calculate fees
-    let mut messages: Vec<&Message> = Vec::new();
-    let (blockhash, fee_calculator, _) = rpc_client
-        .get_recent_blockhash_with_commitment(config.commitment)?
-        .value;
     let minimum_balance = rpc_client.get_minimum_balance_for_rent_exemption(program_data.len())?;
-    let ix = system_instruction::create_account(
-        &config.signers[0].pubkey(),
-        &program_id.pubkey(),
-        minimum_balance.max(1),
-        program_data.len() as u64,
-        &loader_id,
-    );
-    let message = Message::new(&[ix], Some(&config.signers[0].pubkey()));
-    let mut create_account_tx = Transaction::new_unsigned(message);
     let signers = [config.signers[0], program_id];
-    create_account_tx.try_sign(&signers, blockhash)?;
-    messages.push(&create_account_tx.message);
+
+    // Check program account to see if partial initialization has occurred
+    println!("before account check");
+    let initial_instructions = if let Some(account) = rpc_client
+        .get_account_with_commitment(&program_id.pubkey(), config.commitment)?
+        .value
+    {
+        let mut instructions: Vec<Instruction> = vec![];
+        if account.executable {
+            return Err(CliError::DynamicProgramError(
+                "Program account is already executable".to_string(),
+            )
+            .into());
+        }
+        if account.data.is_empty() {
+            instructions.push(system_instruction::allocate(
+                &program_id.pubkey(),
+                program_data.len() as u64,
+            ));
+        }
+        if account.owner != loader_id {
+            instructions.push(system_instruction::assign(&program_id.pubkey(), &loader_id));
+        }
+        if account.lamports < minimum_balance {
+            instructions.push(system_instruction::transfer(
+                &config.signers[0].pubkey(),
+                &program_id.pubkey(),
+                minimum_balance - account.lamports,
+            ));
+        }
```

Comment:
Won't these all occur atomically or be fully rolled back depending on the success of the first initial transaction?  I don't think we can end up with a partially setup account upon a failed `system_instruction::create_account()`

It seems like the queried account will be in one of these states
1. Non-existent - We're starting from scratch
1. Exists, properly allocated, not yet executable - We're retrying deploy
1. Exists, properly allocated, already executable - Nothing to do
1. Exists, otherwise occupied - Will never succeed

Reply:
Yes, they will if you are only using `solana deploy`.

But you can end up with a partially set up account using other methods. The most likely partial setup would be a program account with some SOL already in it -- as easy as `solana transfer`. The other states are much less likely (require doing some subset of the create-account actions independently), but given that the former requires allocate+assign+transfer, I figured might as well be thorough. (And stay in sync with the web3 sdk, for that matter.)

Reply:
Gotcha!  I didn't realize we were trying to support situations that `solana deploy` can't get us in to.  Probably good to have a recovery path for developers who try to be too clever for their own good :wink: 

I don't think allocate or assign will succeed if the system program isn't the owner though.  So maybe toss that extra check in?

Reply:
Fyi, allocate cares, but assign does not: https://github.com/solana-labs/solana/blob/f41a73d76a32faff7f79991fe728d32e38964d19/runtime/src/system_instruction_processor.rs#L86

Reply:
I think assign will hit this post-check?

https://github.com/solana-labs/solana/blob/f41a73d76a32faff7f79991fe728d32e38964d19/runtime/src/message_processor.rs#L85-L91

Reply:
Oh I see, forgot about that check! Updated